src/pallene/pallenelib.lua

```diff
@@ -90,12 +127,209 @@ static lua_Number  pallene_math_log(lua_Integer x, lua_Integer base);
 static lua_Integer pallene_math_modf(lua_State *L, const char* file, int line, lua_Number n, lua_Number* out);
 
 /* Other builtins */
-static TString* pallene_string_char(lua_State *L, const char* file, int line, lua_Integer c);
-static TString* pallene_string_sub(lua_State *L, TString *str, lua_Integer start, lua_Integer end);
+static TString *pallene_string_char(lua_State *L, const char* file, int line, lua_Integer c);
+static TString *pallene_string_sub(lua_State *L, TString *str, lua_Integer start, lua_Integer end);
 static TString *pallene_type_builtin(lua_State *L, TValue v);
 static TString *pallene_tostring(lua_State *L, const char* file, int line, TValue v);
 static void pallene_io_write(lua_State *L, TString *str);
 
+/* Pallene tracer implementation. */
+
+static void pallene_tracer_frameenter(lua_State *L, pt_frame_t * restrict frame, ptrdiff_t sig) {
+    pt_frame_t *head, *tail;
+
+    lua_getglobal(L, "__pallene_tracer_stack_head");
+    head = (pt_frame_t *) lua_topointer(L, -1);
+    lua_getglobal(L, "__pallene_tracer_stack_tail");
+    tail = (pt_frame_t *) lua_topointer(L, -1);
+    lua_pop(L, 2);
+
+    /* If there is no frame in the stack. */
+    if(head == NULL) {
+        /* Just to be safe. */
+        frame -> prev = NULL;
+        frame -> next = NULL;
+
+        /* There is no other signature. */
+        frame -> frame_sig = sig;
+
+        head = frame;
+        tail = frame;
+
+        goto out;
+    }
+
+    /* The frame signature. */
+    /* If we don't have any frame signature, that denotes the call was made from Pallene
+       environment. */
+    frame -> frame_sig = sig == (ptrdiff_t) NULL ? tail -> frame_sig : sig;
+
+    tail  -> next = frame;
+    frame -> prev = tail;
+
+    tail = frame;
+
+out:
+    /* Now update the registry. */
+    lua_pushlightuserdata(L, head);
+    lua_setglobal(L, "__pallene_tracer_stack_head");
+    lua_pushlightuserdata(L, tail);
+    lua_setglobal(L, "__pallene_tracer_stack_tail");
+}
+
+static void pallene_tracer_global_setline(lua_State *L, int line) {
+    lua_getglobal(L, "__pallene_tracer_stack_tail");
+    pt_frame_t *tail = (pt_frame_t *) lua_topointer(L, -1);
+    lua_pop(L, 1);
+
+    if(tail != NULL)
+        tail -> line = line;
+}
+
+static void pallene_tracer_setline(pt_frame_t * restrict frame, int line) {
+    frame -> line = line;
+}
+
+static void pallene_tracer_frameexit(lua_State *L) {
+    pt_frame_t *head, *tail;
+
+    lua_getglobal(L, "__pallene_tracer_stack_head");
+    head = (pt_frame_t *) lua_topointer(L, -1);
+    lua_getglobal(L, "__pallene_tracer_stack_tail");
+    tail = (pt_frame_t *) lua_topointer(L, -1);
+    lua_pop(L, 2);
+
+    /* We are popping the very last frame. */
+    if(tail -> prev == NULL) {
+        tail = NULL;
+        head = NULL;
+
+        goto out;
```

Comment:
I think it'll be easier to follow if we avoid this `goto`.
Maybe take a cursory look at all the `goto`s we have right now and reconsider some of them?

Reply:
`goto`s are extensively used as finalizers. Also used in Linux kernel as well :).

Reply:
Sure, I don't mind them as long as we are consistent with their use.
If you feel this usage is consistent with the code style of the rest of `pallenelib`, we can keep them.